providers/rpc/http.go

```diff
@@ -0,0 +1,77 @@
+package rpc
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// createRequest
+func (c *Config) createRequest(ctx context.Context, p RequestPayload) (*http.Request, error) {
+	var data []byte
+	if rj := c.Opts.Experimental.CustomRequestPayload; rj != nil && c.Opts.Experimental.DotPath != "" {
+		d, err := p.embedPayload(rj, c.Opts.Experimental.DotPath)
+		if err != nil {
+			return nil, err
+		}
+		data = d
+	} else {
+		d, err := json.Marshal(p)
+		if err != nil {
+			return nil, err
+		}
+		data = d
+	}
+
+	req, err := http.NewRequestWithContext(ctx, c.Opts.Request.HTTPMethod, c.listenerURL.String(), bytes.NewReader(data))
+	if err != nil {
+		return nil, err
+	}
+	for k, v := range c.Opts.Request.StaticHeaders {
+		req.Header.Add(k, strings.Join(v, ","))
+	}
+	if c.Opts.Request.HTTPContentType != "" {
+		req.Header.Set("Content-Type", c.Opts.Request.HTTPContentType)
+	}
+	if c.Opts.Request.TimestampHeader != "" {
+		req.Header.Add(c.Opts.Request.TimestampHeader, time.Now().Format(c.Opts.Request.TimestampFormat))
+	}
+
+	return req, nil
+}
+
+func (c *Config) handleResponse(resp *http.Response, reqKeysAndValues []interface{}) (ResponsePayload, error) {
+	kvs := reqKeysAndValues
+	defer func() {
+		if !c.LogNotificationsDisabled {
+			kvs = append(kvs, responseKVS(resp)...)
+			c.Logger.Info("rpc notification details", kvs...)
+		}
+	}()
+	defer resp.Body.Close()
```

Comment:
I don't recall how defer works but are you sure this closes the right handle given we reset the body on l.74?

(it may be the case that Go stores the point in memory it'll jump to on return at the time of `defer`, hence I'm not sure if this is a problem)